From 2c62813d390cd5b9680da0089002708ce134867a Mon Sep 17 00:00:00 2001
From: Sergey Fedorov <barracuda@macos-powerpc.org>
Date: Sun, 11 Oct 2026 08:20:02 +0000
Subject: [PATCH] Revert "Require Mac OS X 10.6 or later"

This reverts commit 3e96f16643af076a923c0adb8fcd5ffc6e1b9bf4.

Part of restoring what https://github.com/ruby/ruby/pull/19183 dropped.

---
 configure.ac     |  6 +++---
 error.c          | 16 ++++++++++++++++
 file.c           | 14 +++++++++++++-
 io.c             | 10 ++++++++++
 signal.c         |  7 ++++++-
 thread_pthread.c | 26 ++++++++++++++++++++++++++
 vm_dump.c        | 10 ++++++++--
 7 files changed, 82 insertions(+), 7 deletions(-)

diff --git a/configure.ac b/configure.ac
index d64e597e81..46648a5c20 100644
--- a/configure.ac
+++ b/configure.ac
@@ -440,9 +440,9 @@ AS_CASE(["$target_os"],
 [darwin*], [
     AC_MSG_CHECKING(if minimum required OS X version is supported)
     AC_PREPROC_IFELSE([AC_LANG_SOURCE([[@%:@include <AvailabilityMacros.h>
-	@%:@if MAC_OS_X_VERSION_MIN_REQUIRED < __MAC_10_6
-	@%:@error pre OS X 10.6
-	[!<===== pre OS X 10.6 =====>]
+	@%:@if MAC_OS_X_VERSION_MIN_REQUIRED < __MAC_10_5
+	@%:@error pre OS X 10.5
+	[!<===== pre OS X 10.5 =====>]
 	@%:@endif
 	]])],
 	[AC_MSG_RESULT(yes)],
diff --git a/error.c b/error.c
index 4e1d96cb9b..0b51fe401c 100644
--- a/error.c
+++ b/error.c
@@ -27,6 +27,10 @@
 # include <sys/wait.h>
 #endif

+#if defined __APPLE__
+# include <AvailabilityMacros.h>
+#endif
+
 #include "internal.h"
 #include "internal/class.h"
 #include "internal/error.h"
@@ -960,6 +964,11 @@ bug_important_message(FILE *out, const char *const msg, size_t len)
     fwrite(p, 1, endmsg - p, out);
 }

+#undef CRASH_REPORTER_MAY_BE_CREATED
+#if defined(__APPLE__) && \
+    (!defined(MAC_OS_X_VERSION_10_6) || MAC_OS_X_VERSION_MIN_REQUIRED < MAC_OS_X_VERSION_10_6)
+# define CRASH_REPORTER_MAY_BE_CREATED
+#endif
 static void
 preface_dump(FILE *out)
 {
@@ -968,6 +977,10 @@ preface_dump(FILE *out)
         "-- Crash Report log information "
         "--------------------------------------------\n"
         "   See Crash Report log file in one of the following locations:\n"
+# ifdef CRASH_REPORTER_MAY_BE_CREATED
+        "     * ~/Library/Logs/CrashReporter\n"
+        "     * /Library/Logs/CrashReporter\n"
+# endif
         "     * ~/Library/Logs/DiagnosticReports\n"
         "     * /Library/Logs/DiagnosticReports\n"
         "   for more details.\n"
@@ -989,6 +1002,9 @@ postscript_dump(FILE *out)
         "[IMPORTANT]"
         /*" ------------------------------------------------"*/
         "\n""Don't forget to include the Crash Report log file under\n"
+# ifdef CRASH_REPORTER_MAY_BE_CREATED
+        "CrashReporter or "
+# endif
         "DiagnosticReports directory in bug reports.\n"
         /*"------------------------------------------------------------\n"*/
         "\n";
diff --git a/file.c b/file.c
index 24d424c9ec..bce6f06322 100644
--- a/file.c
+++ b/file.c
@@ -5651,6 +5651,11 @@ rb_check_realpath_emulate_rescue(VALUE arg, VALUE exc)
 {
     return Qnil;
 }
+#elif !defined(NEEDS_REALPATH_BUFFER) && defined(__APPLE__) && \
+    (!defined(MAC_OS_X_VERSION_10_6) || (MAC_OS_X_VERSION_MIN_REQUIRED < MAC_OS_X_VERSION_10_6))
+/* realpath() on OSX < 10.6 doesn't implement automatic allocation */
+# include <sys/syslimits.h>
+# define NEEDS_REALPATH_BUFFER 1
 #endif /* HAVE_REALPATH */

 static VALUE
@@ -5660,6 +5665,11 @@ rb_check_realpath_internal(VALUE basedir, VALUE path, rb_encoding *origenc, enum
     VALUE unresolved_path;
     char *resolved_ptr = NULL;
     VALUE resolved;
+# if defined(NEEDS_REALPATH_BUFFER) && NEEDS_REALPATH_BUFFER
+    char resolved_buffer[PATH_MAX];
+# else
+    char *const resolved_buffer = NULL;
+# endif

     if (mode == RB_REALPATH_DIR) {
         return rb_check_realpath_emulate(basedir, path, origenc, mode);
@@ -5672,7 +5682,7 @@ rb_check_realpath_internal(VALUE basedir, VALUE path, rb_encoding *origenc, enum
     }
     if (origenc) unresolved_path = TO_OSPATH(unresolved_path);

-    if ((resolved_ptr = realpath(RSTRING_PTR(unresolved_path), NULL)) == NULL) {
+    if ((resolved_ptr = realpath(RSTRING_PTR(unresolved_path), resolved_buffer)) == NULL) {
         /*
            wasi-libc 22 and later support realpath(3) but return ENOTSUP
            when the underlying host syscall returns it.
@@ -5693,7 +5703,9 @@ rb_check_realpath_internal(VALUE basedir, VALUE path, rb_encoding *origenc, enum
         rb_sys_fail_path(unresolved_path);
     }
     resolved = ospath_new(resolved_ptr, strlen(resolved_ptr), rb_filesystem_encoding());
+# if !(defined(NEEDS_REALPATH_BUFFER) && NEEDS_REALPATH_BUFFER)
     free(resolved_ptr);
+# endif

 # if !defined(__linux__) && !defined(__APPLE__)
     /* As `resolved` is a String in the filesystem encoding, no
diff --git a/io.c b/io.c
index c4318906b3..73e0124596 100644
--- a/io.c
+++ b/io.c
@@ -100,6 +100,16 @@

 #ifdef HAVE_COPYFILE_H
 # include <copyfile.h>
+
+# ifndef COPYFILE_STATE_COPIED
+/*
+ * Some OSes (e.g., OSX < 10.6) implement fcopyfile() but not
+ * COPYFILE_STATE_COPIED.  Since the only use of the former here
+ * requires the latter, we disable the former when the latter is undefined.
+ */
+#   undef HAVE_FCOPYFILE
+# endif
+
 #endif

 #if defined __APPLE__
diff --git a/signal.c b/signal.c
index c7ee5acd25..a72283aeaf 100644
--- a/signal.c
+++ b/signal.c
@@ -815,7 +815,12 @@ check_stack_overflow(int sig, const uintptr_t addr, const ucontext_t *ctx)
     const greg_t bp = mctx->gregs[REG_EBP];
 #   endif
 # elif defined __APPLE__
-#   define MCTX_SS_REG(reg) __ss.__##reg
+#   include <AvailabilityMacros.h>
+#   if defined(MAC_OS_X_VERSION_10_5) && MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_5
+#     define MCTX_SS_REG(reg) __ss.__##reg
+#   else
+#     define MCTX_SS_REG(reg) ss.reg
+#   endif
 #   if defined(__LP64__)
     const uintptr_t sp = mctx->MCTX_SS_REG(rsp);
     const uintptr_t bp = mctx->MCTX_SS_REG(rbp);
diff --git a/thread_pthread.c b/thread_pthread.c
index 05e8263c51..3e8d53c3be 100644
--- a/thread_pthread.c
+++ b/thread_pthread.c
@@ -42,6 +42,10 @@

 #include "probes.h"

+#if defined __APPLE__
+# include <AvailabilityMacros.h>
+#endif
+
 #if defined(HAVE_SYS_EVENTFD_H) && defined(HAVE_EVENTFD)
 #  define USE_EVENTFD (1)
 #  include <sys/eventfd.h>
@@ -1121,9 +1125,31 @@ native_thread_native_thread_id(rb_thread_t *target_th)
     return INT2FIX(tid);
 #elif defined(__APPLE__)
     uint64_t tid;
+/* The first condition is needed because MAC_OS_X_VERSION_10_6
+    is not defined on 10.5. Note, 10.5 is supported with GCC upstream,
+    so it has C++17 and everything needed to build modern Ruby. */
+# if (!defined(MAC_OS_X_VERSION_10_6) || \
+      (MAC_OS_X_VERSION_MAX_ALLOWED < MAC_OS_X_VERSION_10_6))
+    const bool no_pthread_threadid_np = true;
+#   define NO_PTHREAD_MACH_THREAD_NP 1
+# elif MAC_OS_X_VERSION_MIN_REQUIRED >= MAC_OS_X_VERSION_10_6
+    const bool no_pthread_threadid_np = false;
+# else
+#   if !(defined(__has_attribute) && __has_attribute(availability))
+    /* __API_AVAILABLE macro does nothing on gcc */
+    __attribute__((weak)) int pthread_threadid_np(pthread_t, uint64_t*);
+#   endif
+    /* Check weakly linked symbol */
+    const bool no_pthread_threadid_np = !&pthread_threadid_np;
+# endif
+    if (no_pthread_threadid_np) {
+        return ULL2NUM(pthread_mach_thread_np(pthread_self()));
+    }
+# ifndef NO_PTHREAD_MACH_THREAD_NP
     int e = pthread_threadid_np(target_th->nt->thread_id, &tid);
     if (e != 0) rb_syserr_fail(e, "pthread_threadid_np");
     return ULL2NUM((unsigned long long)tid);
+# endif
 #endif
 }
 # define USE_NATIVE_THREAD_NATIVE_THREAD_ID 1
diff --git a/vm_dump.c b/vm_dump.c
index 5c6b2cd689..be61114bfb 100644
--- a/vm_dump.c
+++ b/vm_dump.c
@@ -758,7 +758,12 @@ rb_vmdebug_thread_dump_state(FILE *errout, VALUE self)
 }

 #if defined __APPLE__
-# define MCTX_SS_REG(reg) __ss.__##reg
+# include <AvailabilityMacros.h>
+# if defined(MAC_OS_X_VERSION_10_5) && MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_5
+#   define MCTX_SS_REG(reg) __ss.__##reg
+# else
+#   define MCTX_SS_REG(reg) ss.reg
+# endif
 #endif

 #if defined(HAVE_BACKTRACE)
@@ -766,7 +771,8 @@ rb_vmdebug_thread_dump_state(FILE *errout, VALUE self)
 # ifdef HAVE_LIBUNWIND
 #  undef backtrace
 #  define backtrace unw_backtrace
-# elif defined(__APPLE__) && defined(HAVE_LIBUNWIND_H)
+# elif defined(__APPLE__) && defined(HAVE_LIBUNWIND_H) \
+    && defined(MAC_OS_X_VERSION_10_6) && MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_6
 #  define UNW_LOCAL_ONLY
 #  include <libunwind.h>
 #  include <sys/mman.h>
